presto-main/src/main/java/com/facebook/presto/operator/GenericPageProcessor.java

```diff
@@ -36,16 +51,153 @@ public GenericPageProcessor(FilterFunction filterFunction, Iterable<? extends Pr
+                outputBlocks[projectionIndex] = new DictionaryBlock(selectedPositions.length, outputDictionary, wrappedIntArray(outputIds));
+            }
+            else {
+                BlockBuilder blockBuilder = pageBuilder.getBlockBuilder(projectionIndex);
+                for (int position : selectedPositions) {
+                    projection.project(position, inputBlocks, blockBuilder);
+                }
+                outputBlocks[projectionIndex] = blockBuilder.build();
+            }
+        }
+
+        for (Block block : outputBlocks) {
+            verify(block.getPositionCount() == selectedPositions.length);
+        }
+        return new Page(outputBlocks);
+    }
+
+    private static int[] filterIds(ProjectionFunction projection, Page page, int[] selectedPositions)
+    {
+        Slice ids = ((DictionaryBlock) page.getBlock(getOnlyElement(projection.getInputChannels()))).getIds();
+
+        int[] outputIds = new int[selectedPositions.length];
+        for (int pos = 0; pos < selectedPositions.length; pos++) {
+            outputIds[pos] = ids.getInt(selectedPositions[pos] * SizeOf.SIZE_OF_INT);
+        }
+        return outputIds;
+    }
+
+    private static Block projectDictionary(ProjectionFunction projection, Page page)
+    {
+        int inputChannel = getOnlyElement(projection.getInputChannels());
+        Block dictionary = ((DictionaryBlock) page.getBlock(inputChannel)).getDictionary();
+
+        BlockBuilder dictionaryBuilder = projection.getType().createBlockBuilder(new BlockBuilderStatus(), dictionary.getPositionCount());
+        Block[] blocks = new Block[page.getChannelCount()];
+        blocks[inputChannel] = dictionary;
+
+        for (int i = 0; i < dictionary.getPositionCount(); i++) {
+            projection.project(i, blocks, dictionaryBuilder);
+        }
+        return dictionaryBuilder.build();
+    }
+
+    private static boolean canDictionaryProcess(ProjectionFunction projection, Page inputPage)
+    {
+        Set<Integer> inputChannels = projection.getInputChannels();
+        return projection.isDeterministic()
+                && inputChannels.size() == 1
+                && (inputPage.getBlock(getOnlyElement(inputChannels)) instanceof DictionaryBlock);
+    }
+
+    private Page getNonLazyPage(Page page)
+    {
+        ImmutableSet.Builder<Integer> builder = ImmutableSet.builder();
+        for (ProjectionFunction projection : projections) {
+            builder.addAll(projection.getInputChannels());
+        }
+        Set<Integer> inputChannels = builder.build();
+
+        if (inputChannels.isEmpty()) {
+            return page;
+        }
+
+        Block[] blocks = page.getBlocks();
+        for (int inputChannel : inputChannels) {
+            Block block = page.getBlock(inputChannel);
+            if (block instanceof LazyBlock && ((LazyBlock) block).getBlock() instanceof DictionaryBlock) {
```

Comment:
Calling getBlocks forces lazy load, so you should just always replace the block in the page since is avoids the redirection.